neuron/src/lib/Neuron/Zettelkasten/Graph/Build.hs

```diff
@@ -17,20 +17,38 @@ import Neuron.Zettelkasten.ID
 import Neuron.Zettelkasten.Query.Error (QueryError)
 import Neuron.Zettelkasten.Query.Eval (queryConnections)
 import Neuron.Zettelkasten.Zettel
+import Neuron.Zettelkasten.Zettel.Format
 import Neuron.Zettelkasten.Zettel.Parser
 import Relude
+import Relude.Extra.Group
+import System.FilePath
 
 buildZettelkasten ::
-  [(FilePath, Text)] ->
+  [(ZettelFormat, ZettelReader, [(FilePath, Text)])] ->
   ( ZettelGraph,
     [ZettelC],
     Map ZettelID ZettelError
   )
-buildZettelkasten filesWithContent =
-  let zs = parseZettels filesWithContent
+buildZettelkasten filesPerFormat =
+  let allFiles = concatMap (\(_, _, filesWithContent) -> fmap fst filesWithContent) filesPerFormat
+      groupedFiles = groupBy (toText . takeBaseName) allFiles
+      duplicates = Map.filter (\files -> length files > 1) groupedFiles
+      filterUnique (format, zreader, filesWithContent) =
+        ( format,
+          zreader,
+          filter
+            (\(file, _) -> Map.notMember (toText $ takeBaseName file) duplicates)
+            filesWithContent
+        )
```

Comment:
This entire thing should be refactored and made into small function(s).

Also, you should group by the parsed zettel ID, and not base name. Which means that you wouldn't be using `getZettelID` in Zettel/Parser.hs, and instead have `parseZettels` take ZettelID along with FilePath.

Reply:
I now group by ID, which also removes the use of the partial `parseZettelID` function. I also tried splitting the function into smaller parts, but nothing very satisfying yet. Not sure what I can do about it. Maybe use a record instead of tuples everywhere? I have to define a bunch of really basic functions since base/relude usually only care about pairs, here we work with triples.

Reply:
I'll try to refactor it.